Refactor command line arguments

weirddan455 opened the pull request:
Fixes #2121

I didn't end up using `getopt`.  It seemed more trouble than it was worth and didn't fit in well with our existing `CommandLine` class.  It also has some oddities like behaving differently if certain environment variables are set such as `POSIXLY_CORRECT` and it mutates the `argv` argument that it takes in.  There's also the possibility that it behaves differently on different platforms.

Anyway, I think I've cleaned the mess up quite a bit.  All the argument get parsed into a struct in the global Config class on program start.  Then, whenever someone wants to check if an argument is set, they can just directly read the already parsed value.  That way we're not doing parsing stuff all over the code and it should be more maintainable.  If someone wants to add a new argument, it's just adding a variable to the struct and then add a single line to the `ParseArguments()` function.

This mostly removes all the `control->cmdline` calls.  The one place I left that in was Autoexec.  There's a bunch of custom parsing logic in there that I didn't want to touch.

johnnovak commented:
Exciting stuff and long overdue! I'm so glad you've done this @weirddan455, it will be a very nice enhancement for the upcoming release! 🏅 I will review this in detail later today.

> I didn't end up using `getopt`. It seemed more trouble than it was worth and didn't fit in well with our existing `CommandLine` class. It also has some oddities like behaving differently if certain environment variables are set such as `POSIXLY_CORRECT` and it mutates the `argv` argument that it takes in. There's also the possibility that it behaves differently on different platforms.

I think you've done the right thing. I did not research `getopt` in great detail, but definitely the simplest solution that does the job is the correct choice ✅ 

> Anyway, I think I've cleaned the mess up quite a bit. All the argument get parsed into a struct in the global Config class on program start. Then, whenever someone wants to check if an argument is set, they can just directly read the already parsed value. That way we're not doing parsing stuff all over the code and it should be more maintainable. If someone wants to add a new argument, it's just adding a variable to the struct and then add a single line to the `ParseArguments()` function.

Sounds awesome! Parsing stuff in random places all over the codebase was utter crap, yeah... So glad that has been cleaned up!

> This mostly removes all the `control->cmdline` calls. The one place I left that in was Autoexec. There's a bunch of custom parsing logic in there that I didn't want to touch.

Good stuff!

johnnovak commented on include/control.h:
```diff
@@ -40,15 +40,42 @@ enum class Verbosity : int8_t {
+	std::vector<std::string> conf;
+	std::vector<std::string> set;
+	std::optional<std::vector<std::string>> editconf;
+	std::optional<int> socket;
```
It's a good idea to keep using the command line switch names here for clarity 👍🏻

johnnovak commented on src/gui/sdlmain.cpp:
```diff
@@ -5004,7 +4995,7 @@ int sdl_main(int argc, char *argv[])
+	} catch (const std::exception& e) {
 		// catch all exceptions that derive from the standard library
 		LOG_ERR("EXCEPTION: Standard library exception: %s", e.what());
 		rcode = 1;
```
Everything reads better and is better organise this way, well done 🎉

johnnovak commented on include/programs.h:
```diff
@@ -58,14 +61,22 @@ class CommandLine {
 	                   const std::list<std::string_view>& post_args) const;
 	void Shift(unsigned int amount = 1);
 	uint16_t Get_arglength();
+	bool FindBoolArgument(const std::string& name, bool remove,
```
I'd introduce two version of this for consistency with the other functions, and extra boolean args are best to be avoided:

- `FindRemoveBoolArgument(const std::string& name, char short_letter = 0)`
- `FindBoolArgument(const std::string& name, char short_letter = 0)`

weirddan455 replied:
Fixed

johnnovak commented on src/misc/setup.cpp:
```diff
@@ -1789,3 +1790,35 @@ const char* SetProp(std::vector<std::string>& pvars)
+	                        cmdline->FindBoolArgument("resetmapper", true);
+
+	arguments.version = cmdline->FindBoolArgument("version", true, 'v');
+	arguments.help    = cmdline->FindBoolArgument("help", true, 'h');
```
As I mentioned above, `FindRemoveBoolArgument("help", 'h')` etc. would be much nicer. Second best would be:

```cpp
constexpr auto remove_arg = true;
...
FindBoolArgument("help", remove_arg, 'h');
...
```

weirddan455 replied:
I did that 2nd approach just to avoid duplicated code.  I made `FindBoolArgument` a private method because it was only used in once place inside the `CommandLine` class.

johnnovak commented on src/shell/command_line.cpp:
```diff
@@ -374,3 +374,74 @@ void CommandLine::Shift(unsigned int amount)
+	short_name[1]                 = short_letter;
+	return FindExist(double_dash.c_str(), remove) ||
+	       FindExist(dash.c_str(), remove) ||
+	       (short_letter && FindExist(short_name, true));
```
`FindExist(short_name, true)` — what is `true`? Suggest extracting it to a const with descriptive name.

weirddan455 replied:
`true` was a bug in this case.  It was supposed to be the `remove` flag.  Fixed.

johnnovak replied:
Yeah in multi-arg functions it's almost never apparent what something like this means: `foo(blah, 4, 0, true, 233, false, false)` 🥶

johnnovak commented on src/shell/command_line.cpp:
```diff
@@ -374,3 +374,74 @@ void CommandLine::Shift(unsigned int amount)
+std::string CommandLine::StringArgHelper(const char* name)
+{
+	cmd_it it;
+	while (FindEntry(name, it, true)) {
```
Extract `true`.

weirddan455 replied:
Fixed

johnnovak commented on src/shell/command_line.cpp:
```diff
@@ -374,3 +374,74 @@ void CommandLine::Shift(unsigned int amount)
+	       (short_letter && FindExist(short_name, true));
+}
+
+std::string CommandLine::StringArgHelper(const char* name)
```
How does it help exactly? 😄 Can you try to make the name reflect what the function is actually doing? "Helper", "Util", etc. can be literally anything.

weirddan455 replied:
Renamed to `FindRemoveSingleString`

johnnovak commented on src/shell/command_line.cpp:
```diff
@@ -374,3 +374,74 @@ void CommandLine::Shift(unsigned int amount)
+	const std::string double_dash = "--" + name;
+	const std::string dash        = '-' + name;
+
+	std::string ret = StringArgHelper(double_dash.c_str());
```
Yeah, I'm not sure what the help is 😆

johnnovak replied:
`const some_better_name = ...`

weirddan455 replied:
Renamed to `FindRemoveSingleString`

johnnovak commented on src/shell/command_line.cpp:
```diff
@@ -374,3 +374,74 @@ void CommandLine::Shift(unsigned int amount)
+
+	std::string ret = StringArgHelper(double_dash.c_str());
+	if (ret.empty()) {
+		ret = StringArgHelper(dash.c_str());
```
Just a minor style comment: return from here, then you don't need to mutate `ret`.

weirddan455 replied:
Fixed

johnnovak commented on src/shell/command_line.cpp:
```diff
@@ -374,3 +374,74 @@ void CommandLine::Shift(unsigned int amount)
+std::vector<std::string> CommandLine::FindRemoveVectorArgument(const std::string& name)
+{
+	std::vector<std::string> ret = {};
+	while (1) {
```
Minor: i `for (;;)` is pretty standard, or even `while (true)`. Kinda dislike there's a number there... weird, i know 😄

johnnovak commented on src/shell/command_line.cpp:
```diff
@@ -374,3 +374,74 @@ void CommandLine::Shift(unsigned int amount)
+
+std::vector<std::string> CommandLine::FindRemoveVectorArgument(const std::string& name)
+{
+	std::vector<std::string> ret = {};
```
Maybe use it `arg`. I don't really like the generic `ret` name in functions.

weirddan455 replied:
Fixed

johnnovak commented on src/shell/command_line.cpp:
```diff
@@ -374,3 +374,74 @@ void CommandLine::Shift(unsigned int amount)
+std::optional<std::vector<std::string>> CommandLine::FindRemoveOptionalArgument(
+        const std::string& name)
+{
+	if (!FindBoolArgument(name, false)) {
```
What's false?

weirddan455 replied:
Extracted the value to `remove_arg`

johnnovak commented:
Very nice @weirddan455. Gave it some quick testing, works well.

Next thing would be updating the actual help text in `gui_msgs.h` so it's consistent, lists all the supported command line options (not just the "common options"...), and it's not a weird mismash of single and double-dashes 🤦🏻

All you up for the task? 😎

johnnovak commented on src/shell/command_line.cpp:
```diff
@@ -374,3 +374,74 @@ void CommandLine::Shift(unsigned int amount)
+
+std::string CommandLine::StringArgHelper(const char* name)
+{
+	cmd_it it;
```
Best to init with `{}`, just to be consistent.

johnnovak commented:
This is looking good now @weirddan455 . So how about updating the actual help text in `gui_msgs.h`?

weirddan455 commented:
> This is looking good now @weirddan455 . So how about updating the actual help text in `gui_msgs.h`?

Sure, I'll get to that in a little bit.

weirddan455 commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
 
-  -nolocalconf         Don't read settings from "dosbox.conf" if present in
-                       the current working directory.
+  --securemode             Start in secure mode.
```
I don't know what secure mode does exactly.  Can you help with a description here?

johnnovak replied:
From the DOSBox X manual:

> Enables DOSBox-X’s secure mode. The [config] and [autoexec] sections of the loaded configuration file will be skipped, and commands such as MOUNT and IMGMOUNT are disabled.

https://github.com/joncampbell123/dosbox-x/wiki/DOSBox%E2%80%90X%E2%80%99s-Command%E2%80%90Line-Options

You can use this with a few changes.

weirddan455 replied:
Good stuff.  I removed the DOSBox-X name and used that.

weirddan455 commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
 
-  --working-dir <path> Set working directory to <path>. DOSBox will act as if
-                       started from this directory.
+  --opencaptures           Launches some external program (I don't even know...)
```
This runs the following function (note the TODO comment).  It looks like the user is supposed to specify an external program they want to run against the captures directory.  Not sure what the purpose of this is.

https://github.com/dosbox-staging/dosbox-staging/blob/1930f7f3cd92d2615b5e65059c2ac7350ca85475/src/gui/sdlmain.cpp#L4683-L4706

johnnovak replied:
It' a bizarre feature... Let's just say the same thing as the X manual:

> Calls program with as first parameter the location of the captures folder.

weirddan455 replied:
Sounds good.

weirddan455 commented:
I updated the help with what I could.  Need some help on descriptions for the two above arguments.  I set them all to have double-dashes for long arguments and single dashes for single letter arguments, which is kind of standard.  We support both variants for either type of argument but I think duplicating that in help text is wasted space IMO (could change if you have strong feelings on it).

I also extracted a couple of arguments from `autoexec.cpp`

johnnovak commented:
> We support both variants for either type of argument but I think duplicating that in help text is wasted space IMO (could change if you have strong feelings on it).

Oh yeah, single dashes for long options are only for backward compatibility, but their usage should not be encouraged going forward.

I'll help with those two options later.

johnnovak commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
 
-  -c <command>         Run the specified DOS command before running FILE.
-                       Multiple commands can be specified.
+  --eraseconf              Delete the configuration file.
```
We need to specify which one (default or local).

weirddan455 replied:
It does `const auto path = (get_platform_config_dir() / file).string();` so the default one.  I'll specify.

weirddan455 replied:
Specified "default"

johnnovak commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
+  --eraseconf              Delete the configuration file.
 
-  -noautoexec          Don't perform any [autoexec] actions.
+  --erasemapper            Delete the mapper file.
```
I assume the one defined in the default config? Or whatever `mapperfile` points to after loading the config? Can we make this clearer, I'd be afraid to use it for the first time going by this description.

weirddan455 replied:
This one is also the default `const auto path = (get_platform_config_dir() / MAPPERFILE).string();`

weirddan455 replied:
Specified "default"

johnnovak commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
+                           Example: --set mididevice=fluidsynth
 
-  -lang <langfile>     Start with the language specified in <langfile>.
+  --socket <num>           Runs nullmodem on the specified socket number
```
You're using imperative verbs consistently everywhere else (which is good), so this should be "run".

johnnovak replied:
Please put a period at the end of every config description.

weirddan455 replied:
Fixed

johnnovak commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
+  --opencaptures           Launches some external program (I don't even know...)
 
-  -fullscreen          Start in fullscreen mode.
+  --set <property>=<value> Sets a configuration property.
```
"Set" (not "sets").

johnnovak replied:
We should mention multiple `--set` options can be provided.

weirddan455 replied:
Well, each one needs to be prefixed with `--set`.  Although it kind of works by accident if you don't.  For example, `--set mididevice=fluidsynth soundfont=font.sf2`.  `soundfont=font.sf2` gets sent to DOS which can also set these options.

weirddan455 replied:
I added multiples to the example to clarify that.

johnnovak commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
-                       Valid choices are: hercules, cga, cga_mono, tandy,
-                       pcjr, ega, svga_s3 (default), svga_et3000, svga_et4000,
-                       svga_paradise, vesa_nolfb, vesa_oldvbe.
+  --noautoexec             Don't perform any [autoexec] actions.
```
"Don't execute" would be better.

johnnovak replied:
Also, you've listed this twice; delete the previous one.

weirddan455 replied:
Fixed

johnnovak commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
+
+  -h, --help               Print this help message and exit.
+
+  -v, --version            Print version information and exit.
```
Can we remove the two lines below? The man page is either going away or it will contain the same info, same for the README, plus that path is distro/installations specific. Better to just rid of them.

weirddan455 replied:
Yeah, I was eyeing those lines suspiciously.  Removed.

johnnovak commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
 List of common options:
 
-  --printconf          Print the location of the default configuration file.
+  --printconf              Print the location of the default configuration file.
```
Change the above to "Available options" or something similar (we are aiming to list everything here).

johnnovak replied:
About `--printconf`, hold on to your hat! This also _writes_ the default config if it does not exist 😐 Well, that's how it works, so we need to document it...

weirddan455 replied:
Changed to "available"

weirddan455 replied:
> About `--printconf`, hold on to your hat! This also _writes_ the default config if it does not exist neutral_face Well, that's how it works, so we need to document it...

So it does.  That's unexpected.  Documented.

johnnovak commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
+                           pcjr, ega, svga_s3 (default), svga_et3000, svga_et4000,
+                           svga_paradise, vesa_nolfb, vesa_oldvbe.
+
+  --exit                   Exit after the DOS program specified by FILE has ended.
```
Doesn't this work without specifying FILE, in which case DOSBox exits after executing the autoexec sections? Can you test it out and modify accordingly?

weirddan455 replied:
That is correct.  Added a clarifying line.

johnnovak commented on src/gui/gui_msgs.h:
```diff
@@ -37,45 +37,62 @@ constexpr char help_msg[] =
+                           the current working directory.
 
-  -v, --version        Print version information and exit.
+  --conf <configfile>      Start with the options specified in <configfile>.
```
Would better to say multiple configs can be specified; by multiple args to a single `--conf` option, or by multiple `--conf` options? Can you test it out and clarify?

weirddan455 replied:
It needs a `--conf` before each file.

weirddan455 replied:
Added an example to clarify how to set multiple conf files.

weirddan455 commented:
Does that look good?  I'll squash it down before you merge.  Maybe 2 commits, one for the refactoring and one for the help changes?

I usually like to do the squash at the end to make reviewing easier.

johnnovak commented:
> Does that look good? I'll squash it down before you merge. Maybe 2 commits, one for the refactoring and one for the help changes?
>
> I usually like to do the squash at the end to make reviewing easier.

Thanks for addressing the comments so quickly @weirddan455, it all looks excellent now. 🎉 Great job! 🚀 

Suggest squashing this into 3 commits: refactoring, actual code changes, and updating the help text.

---

And one last suggestion, it would be better to group the options loosely by function and list them in rough order of importance. It's somewhat subjective, but more frequently used options should come earlier.

This is my stab at it; I think it's an improvement. if you agree with it, you can basically copy/paste it (it contains no tabs, only spaces). I've also wrapped a few long lines to make them a bit shorter.

```
Usage: dosbox [OPTION]... [FILE]

List of available options:

  --conf <configfile>      Start with the options specified in <configfile>.
                           Multiple configfiles can be specified.
                           Example: --conf conf1.conf --conf conf2.conf

  --printconf              Print the location of the default configuration file.
                           If it does not exist, create it.

  --editconf               Open the default configuration file in a text editor.

  --eraseconf              Delete the default configuration file.

  --noprimaryconf          Don't read settings from the primary configuration file
                           located in your user folder.

  --nolocalconf            Don't read settings from "dosbox.conf" if present in
                           the current working directory.

  --set <property>=<value> Set a configuration property.
                           Example: --set mididevice=fluidsynth --set soundfont=mysoundfont.sf2

  --working-dir <path>     Set working directory to <path>. DOSBox will act as if
                           started from this directory.

  --list-glshaders         List available GLSL shaders and their directories.
                           Results are useable in the "glshader = " config setting.

  --fullscreen             Start in fullscreen mode.

  --lang <langfile>        Start with the language specified in <langfile>.

  --machine <type>         Emulate a specific type of machine. The machine type has
                           influence on both the emulated video and sound cards.
                           Valid choices are: hercules, cga, cga_mono, tandy,
                           pcjr, ega, svga_s3 (default), svga_et3000, svga_et4000,
                           svga_paradise, vesa_nolfb, vesa_oldvbe.

  -c <command>             Run the specified DOS command before running FILE.
                           Multiple commands can be specified.

  --noautoexec             Don't execute any [autoexec] actions.

  --exit                   Exit after the DOS program specified by FILE has ended.
                           If no FILE has been specified, execute [autoexcec] and
                           terminate.

  --startmapper            Run the mapper GUI.

  --erasemapper            Delete the default mapper file.

  --securemode             Enable secure mode. The [config] and [autoexec] sections
                           of the loaded configuration file will be skipped,
                           and commands such as MOUNT and IMGMOUNT are disabled.

  --socket <num>           Run nullmodem on the specified socket number.

  --opencaptures           Call program with as first parameter the location of the
                           captures folder.

  -h, --help               Print this help message and exit.

  -v, --version            Print version information and exit.
```

weirddan455 commented:
That help text looks good to me.  Thanks for helping out with that :smile: 

>  Suggest squashing this into 3 commits: refactoring, actual code changes, and updating the help text.

I wasn't sure where to draw the line between refactoring and code changes.  I originally just had that as one commit.  I don't believe I changed any behavior except for accepting both single and double dash arguments everywhere and fixing a bug where some arguments were not removed leading to them getting passed to DOS.  So "actual code changes" should be minimal?

Anyway, I squashed it down to 2 commits and rebased to latest main to make sure there weren't any merge conflicts (there weren't :smile: ).  If you can point to where in that first commit should be split up, I can try and do that if you like.

weirddan455 commented:
PVS is failing on main.  That was passing on this PR before I rebased.

FeralChild64 commented:
> PVS is failing on main. That was passing on this PR before I rebased.

PR https://github.com/dosbox-staging/dosbox-staging/pull/2715 will fix it.

kcgen commented:
> PVS is failing on main. That was passing on this PR before I rebased.

:+1: - @johnnovak and @FeralChild64 are on it! (cleanup PR: https://github.com/dosbox-staging/dosbox-staging/pull/2715 will get back down [from 275 to 270 warnings](https://github.com/dosbox-staging/dosbox-staging/actions/workflows/pvs-studio.yml?query=branch%3Amain))

It looks like this branch's count here is 276, so I _think_ there's just one new issue lurking in the changes! if you can ferret it out (or wait and rebase once #2715 goes in)

weirddan455 commented:
>  It looks like this branch's count here is 276, so I think there's just one new issue lurking in the changes! 

How can you tell what the issues actually are?  I see the breakdown of number of issues per file but I don't see what the warnings/errors actually are.

FeralChild64 commented:
Click _Summary_ (left side), download `pvs-analysis-report` file, unpack it, open the report with a web browser.
Indeed, getting the actual warnings is not straightforward.

weirddan455 commented:
> Click _Summary_ (left side), download `pvs-analysis-report` file, unpack it, open the report with a web browser. Indeed, getting the actual warnings is not straightforward.

Thanks.  I found this in a couple places in changes I made in this PR.  It's labeled micro-optimization.  `Decreased performance. Consider creating a reference to avoid using the 'control->arguments' expression repeatedly.`

weirddan455 commented:
Looks like the Mac builds are having a fit now (unrelated to this PR):

```
 ==> Pouring openssl@3--3.1.2.ventura.bottle.tar.gz
Error: Could not symlink bin/openssl
Target /usr/local/bin/openssl
is a symlink belonging to openssl@1.1. You can unlink it:
  brew unlink openssl@1.1

To force the link and overwrite all conflicting files:
  brew link --overwrite openssl@3

To list all files that would be deleted:
  brew link --overwrite --dry-run openssl@3
Error: Process completed with exit code 1.
```

OpenSSL 3.1.2 was just released yesterday.  These CI machines constantly updating packages and breaking is getting annoying...

weirddan455 commented:
Anyway, I pushed a commit to address the PVS warnings.  It's down to 273 now so maybe I inadvertently fixed some other warnings along the way.  I can rebase once the other PR gets merged if you want the green check.

kcgen commented:
Nice @weirddan455 -> yup, suggest letting the other PR reconcile first then rebase to get the new totals.

(I'm guessing you've reduced the total even more, and you can tighten down the limit lower than 270 :partying_face: !).

johnnovak commented:
> That help text looks good to me. Thanks for helping out with that 😄

Teamwork, man 😎 

> Anyway, I squashed it down to 2 commits and rebased to latest main to make sure there weren't any merge conflicts (there weren't 😄 ). If you can point to where in that first commit should be split up, I can try and do that if you like.

Yeah, I haven't double-checked if separating the cleanup/reformatting makes sense or not in this. Sometimes in bigger or more sensitive jobs it helps to do the cleanup/reformat upfront (especially the reformat), so we can see in subsequent commits the actual changes. Not an issue in this case, so the two commits you create are fine (you can rebase the PVS studio stuff into the first one). Just use your judgment in the future.

johnnovak commented:
Main is supposed to be fixed, rebase away @weirddan455 🚀

weirddan455 commented:
I just pushed a rebase.  These Mac builds are still failing with that OpenSSL error.  I definitely rebased to latest main that includes the fix @kcgen made so I'm not sure what's happening :cry:

kcgen commented:
> I just pushed a rebase. These Mac builds are still failing with that OpenSSL error. I definitely rebased to latest main that includes the fix @kcgen made so I'm not sure what's happening cry

This could be GitHub feeding us different VMs; this sometimes that happens when they migrate configurations. Trying a re-run right now; worst-case we'll forcefully remove openssl 1.x prior to the install.

kcgen commented:
Yup; still a problem. Unlink added. can you rebase against `main`?
(sorry for these CI troubles, @weirddan455 !)

weirddan455 commented:
@kcgen I just rebased again... Mac builds still failing :sob:

kcgen commented:
Ok this is really confusing.

`main` is OK but this branch is still haven't problems even after rebasing.

![2023-08-02_19-07](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/76aee6dc-d124-4068-8bb2-ec3451450c52)

![2023-08-02_19-07-2](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/75a2669b-fd1e-4076-920c-21c6521abf3e)

kcgen commented:
@weirddan455 , ok, let's try even harder.

In your branch, try:

```
git cherry-pick d0accd729
```

Then push.  If it works, then we're good.  If not, you can drop/delete that commit.

weirddan455 commented:
Everything is green now!  Thanks @kcgen for emerging victorious over these strange build system issues.

johnnovak commented:
Thanks for persevering @weirddan455 👍🏻

Grounded0 commented:
Connected issues and PRs:

- #2837
